File: src/go/problems/length_of_longest_substring_test.go
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type LengthOfLongestSubstring struct{}

func TestLengthOfLongestSubstring(t *testing.T) {
	gen := core.TestSuite[LengthOfLongestSubstring]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("abcabcbb").Result(3)
	}).Add(func(tc *core.TestCase) {
		tc.Param("bbbbb").Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param("pwwkew").Result(3)
	}).Run(t)
}

func (LengthOfLongestSubstring) Solution(s string) int {
	set := map[byte]struct{}{}

	res := 0
	left := 0

	for right := 0; right < len(s); right++ {
		if _, ok := set[s[right]]; ok {
			res = max(res, right-left)
			for {
				if _, ok := set[s[right]]; !ok {
					break
				}
				delete(set, s[left])
				left++
			}
		}

		set[s[right]] = struct{}{}
	}

	return max(res, len(s)-left)
}

File: src/python/problems/length_of_longest_substring_test.py
#https://leetcode.com/problems/longest-substring-without-repeating-characters/

from core.problem_base import *

class LengthOfLongestSubstring(ProblemBase):
    def Solution(self, s: str) -> int:
        map = {}

        result = 0
        left = 0
        for right in range(len(s)):
            if s[right] not in map:
                result = max(result, right - left + 1)  
            else: 
                if map[s[right]] < left:
                    result = max(result, right - left + 1)
                else: 
                    left = map[s[right]] + 1
            map[s[right]] = right
                
        return result

if __name__ == '__main__':
    TestGen(LengthOfLongestSubstring) \
        .Add(lambda tc: tc.Param("abcabcbb").Result(3)) \
        .Add(lambda tc: tc.Param("bbbbb").Result(1)) \
        .Add(lambda tc: tc.Param("pwwkew").Result(3)) \
        .Run()
